### ISSUE TITLE
Establish clear PR review guideline and document it

### ISSUE THREAD

**Issue by humphd** [mozilla/thimble.mozilla.org#1873]
In 2017 we have a very active contributor and localization community:

![screen shot 2017-03-13 at 11 25 00 am](https://cloud.githubusercontent.com/assets/427398/23861203/bfed4c04-07df-11e7-8905-84e82522a943.png)

I'd like to continue this trend, and part of doing that is making sure we're reviewing and landing code as fast as our developers are writing it.  Right now, for example, there are more than 30 PRs waiting on review.

One of the things that I think is breaking down is that none of us knows exactly where the responsibility lies.  Who *needs* to review certain things?  Who *will* land a patch?  What's our preferred way for landing (just merge, rebase and squash, etc).

Let's establish how we want to do this in the bug, then document it in some wiki page or contributing guide.  Then let's follow it :)

Here's my initial proposal:

* Every PR needs review.  Ideally, we should use Github's new **Reviewers** feature, and someone should set who is on the hook for reviewing (could be 1 or more people, depending on the below).
* A review has multiple parts that need to be addressed, including:
    * Someone needs to actually test it (pull the branch locally, run it, see if it does what it says it does).  As far as I'm concerned, this could be anyone in our community--we should distribute this load.
    * Someone needs to look at the commit for "landability" (e.g., what needs to happen wrt git)
    * Someone needs to look at the code, is it correct, secure, "clean", etc.
    * If it touches UX, someone needs to figure out if it follows our visual guidelines
    * If it touches any strings, someone needs to make sure we haven't broken l10n
    * etc.
* All of the above takes time, and we need to express this to the community to stop the constant thread of "please review this now" type emails and pings.  If it takes 1-2 weeks do do a review, so be it.  We just have to communicate that so people have the right expectations.
* Once a review is done, who is responsible to land it?  The reviewer?  Someone else?
* What is the process for having things get shipped to prod?  We auto-land on staging now, which is great.  When do we update prod?  How is that decided and by whom?

cc @Pomax, @gideonthomas, @flukeout, @ryanwarsaw, @TheoChevalier, who have been active in dealing with issues related to the above in the past, and likely have some insights.

Let's try to agree on something ASAP, write it up, and start doing it.

P.S. I'm also cc'ing some friends who I'd love to see still work with us again: @alicoding, @sedge, @k88hudson, @jbuck.  I realize you are all busy with other things, but if you're ever looking to get involved at all in a fun open source project, consider making it Thimble :)

**Comment by alicoding** [comment]
I'm down helping review whenever needed (when I'm not swamped with work) and would love to set process in term of what would be the acceptable to merge pull request. I also like the idea of using Github's new Reviewers feature.

Question about our Tests here, do we have it cover to at least rely on Travis to help us reduce any error of merging code that we "thought it looks good" ?

**Comment by humphd** [comment]
@alicoding, thanks for jumping in, I'd love to see your expertise still stay connected to this work, even if you're just in an advisement role.

Our testing is very limited at the moment.  We do use Travis for Thimble and Brackets to catch linting errors, but that's about it.  Adding some better test coverage would help with this issue, but it will also distract us.  There's likely a good balance to be had.

#1606 is related to this.

**Comment by Pomax** [comment]
One thing I'd like to suggest is that we revive an issue/PR triage, in the old Thimble style, where we run through PRs first, assign reviewers without discussing the PR itself, and then move on to issues, where we assign a driver, again without discussing the issues itself.

**Comment by humphd** [comment]
@Pomax I think that's a great idea, and it assumes we're doing weekly triage.  Let's commit to all of that.

**Comment by ryanwarsaw** [comment]
There are definitely a lot of basic area's of improvement we could benefit from, including more thorough documentation, improved unit testing, systems and processes in places for review, etc.

I'm a big fan of the two-person system for reviews, and I'd encourage using the trusted contributor hierarchy for initial reviews, but always requiring sign-off by a staff or at least another individual.

One main issue I've noticed is that it's really hard to follow along with the progress of a PR, since the communication is happening on the issue itself, in addition to the PR, Mattermost, Email, etc.

Providing a 1-2 week timeframe for when a PR will be reviewed is way too long, acknowledgement should be within the day and at the maximum it should only take a couple days for review. We want to keep contributors engaged, and this would be counter-productive in that respect. 

Better issue prioritization and identification is important, issues regarding project growth and health, documentation and testing should all be critical issues. These are important to recognize and prioritize, that's something I've particularly noticed with repeat issues during setup, etc.

We really gotta get rid of the (R+/R-/R?) system, even I find myself guilty of doing it but the reality is contributors have no clue what this means and that has a negative effect on message transfer. 

I'm a big fan of the OLTS program and the methods detailed within it for working open. I think we would benefit greatly from implementing those methods here, you can read the full guide here: https://mozilla.github.io/open-leadership-training-series/. 

I like the concept of having a system of defining clearly who is responsible for reviewing what, a lot of open source projects (particularly at Google) have a method of doing this through file change analysis and an automated Github bot that'll go through and assign people.

Sorry this is kinda long, just wanted to get some stuff out there.

**Comment by TheoChevalier** [comment]
> If it touches any strings, someone needs to make sure we haven't broken l10n

Would be happy to provide reviews for this part

**Comment by flukeout** [comment]
Here's the Git collaboration workflow bit that @ryanwarsaw is referencing:
https://mozilla.github.io/open-leadership-training-series/articles/github-for-collaboration/collaborative-workflow-for-contributions/

**Comment by humphd** [comment]
Going to close this, but first a note about what happened during this last term:

* 101 contributors
* 795 commits (557 Thimble + 283 Brackets)
* 486 bugs worked on
* 279 PRs
* 100K lines of code added/removed

Managing that with such a small team of mentors was...a lot!  In the fall I'll be doing it again, we should ponder ways to make it even better.

Thank you to all of you who got so involved in the process.
